rclcpp/include/rclcpp/parameter.hpp

```diff
@@ -0,0 +1,225 @@
+  {
+    value_.integer_value = int_value;
+    value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_INTEGER;
+  }
+  ParameterVariant(const std::string & name, const double double_value)
+  : name_(name)
+  {
+    value_.double_value = double_value;
+    value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_DOUBLE;
+  }
+  ParameterVariant(const std::string & name, const std::string & string_value)
+  : name_(name)
+  {
+    value_.string_value = string_value;
+    value_.parameter_type = rcl_interfaces::ParameterType::PARAMETER_STRING;
+  }
+
+  /* Templated getter */
+  template<typename T>
+  T
+  get_value() const;
+
+  inline ParameterType get_typeID() const {return static_cast<ParameterType>(value_.parameter_type); }
+
+  inline std::string get_name() const {return name_; }
+
+  inline rcl_interfaces::ParameterValue get_parameter_value()
+  {
+    return value_;
+  }
+
+private:
+  std::string name_;
+  rcl_interfaces::ParameterValue value_;
+};
+
+template<>
+inline int64_t ParameterVariant::get_value() const
+{
+  if (value_.parameter_type != rcl_interfaces::ParameterType::PARAMETER_INTEGER) {
+    // TODO: use custom exception
+    throw std::runtime_error("Invalid type");
+  }
+  return value_.integer_value;
+}
+template<>
+inline double ParameterVariant::get_value() const
+{
+  if (value_.parameter_type != rcl_interfaces::ParameterType::PARAMETER_DOUBLE) {
+    // TODO: use custom exception
+    throw std::runtime_error("Invalid type");
+  }
+  return value_.double_value;
+}
+template<>
+inline std::string ParameterVariant::get_value() const
+{
+  if (value_.parameter_type != rcl_interfaces::ParameterType::PARAMETER_STRING) {
+    // TODO: use custom exception
+    throw std::runtime_error("Invalid type");
+  }
+  return value_.string_value;
+}
+template<>
+inline bool ParameterVariant::get_value() const
+{
+  if (value_.parameter_type != rcl_interfaces::ParameterType::PARAMETER_BOOL) {
+    // TODO: use custom exception
+    throw std::runtime_error("Invalid type");
+  }
+  return value_.bool_value;
+}
+
+class AsyncParametersClient
+{
+
+public:
+  AsyncParametersClient(rclcpp::node::Node::SharedPtr node)
+  : node_(node)
+  {
+    get_parameters_client_ = node_->create_client<rcl_interfaces::GetParameters>(
+      "get_parameters");
+    get_parameter_types_client_ = node_->create_client<rcl_interfaces::GetParameterTypes>(
+      "get_parameter_types");
+    set_parameters_client_ = node_->create_client<rcl_interfaces::SetParameters>(
+      "set_parameters");
+    list_parameters_client_ = node_->create_client<rcl_interfaces::ListParameters>(
+      "list_parameters");
+    describe_parameters_client_ = node_->create_client<rcl_interfaces::DescribeParameters>(
+      "describe_parameters");
+  }
+
+  std::shared_future<std::vector<ParameterVariant>>
+  get_parameters(
+    std::vector<std::string> names,
+    std::function<void(
+      std::shared_future<std::vector<ParameterVariant>>)> callback = nullptr)
```

Comment:
This signature left me with some questions, which might just be solved with documentation:

- Why does the callback receive the future and not just the value?
 - I guess because the future could result in an exception?
- Is the expectation that the call to `future.get()` in the user's callback will either raise or return the value without blocking (is the future always ready when the callback is called)?
 - Put another way, can the call to `future.get()` block in the user's callback?
- What happens if the user doesn't call `.get()` on the future?